oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticBulkProcessorHandler.java

```diff
@@ -0,0 +1,296 @@
+     * This option is available for sync index definitions only.
+     */
+    public static ElasticBulkProcessorHandler getBulkProcessorHandler(@NotNull ElasticConnection elasticConnection,
+                                                                      @NotNull ElasticIndexDefinition indexDefinition,
+                                                                      @NotNull NodeBuilder definitionBuilder, CommitInfo commitInfo) {
+        PropertyState async = indexDefinition.getDefinitionNodeState().getProperty("async");
+
+        if (async != null) {
+            return new ElasticBulkProcessorHandler(elasticConnection, indexDefinition, definitionBuilder);
+        }
+
+        // commit-info has priority over configuration in index definition
+        String syncMode = null;
+        if (commitInfo != null) {
+            syncMode = (String) commitInfo.getInfo().get(SYNC_MODE_PROPERTY);
+        }
+
+        if (syncMode == null) {
+            PropertyState syncModeProp = indexDefinition.getDefinitionNodeState().getProperty("sync-mode");
+            if (syncModeProp != null) {
+                syncMode = syncModeProp.getValue(Type.STRING);
+            }
+        }
+
+        if (SYNC_RT_MODE.equals(syncMode)) {
+            return new RealTimeBulkProcessorHandler(elasticConnection, indexDefinition, definitionBuilder);
+        }
+
+        return new ElasticBulkProcessorHandler(elasticConnection, indexDefinition, definitionBuilder);
+    }
+
+    private BulkProcessor initBulkProcessor() {
+        return BulkProcessor.builder(requestConsumer(),
+                new OakBulkProcessorListener())
+                .setBulkActions(indexDefinition.bulkActions)
+                .setBulkSize(new ByteSizeValue(indexDefinition.bulkSizeBytes))
+                .setFlushInterval(TimeValue.timeValueMillis(indexDefinition.bulkFlushIntervalMs))
+                .setBackoffPolicy(BackoffPolicy.exponentialBackoff(
+                        TimeValue.timeValueMillis(indexDefinition.bulkRetriesBackoff), indexDefinition.bulkRetries)
+                )
+                .build();
+    }
+
+    protected BiConsumer<BulkRequest, ActionListener<BulkResponse>> requestConsumer() {
+        return (request, bulkListener) -> elasticConnection.getClient().bulkAsync(request, RequestOptions.DEFAULT, bulkListener);
+    }
+
+    public void add(DocWriteRequest<?> request) {
+        bulkProcessor.add(request);
+        totalOperations++;
+    }
+
+    public boolean close() {
+        if (totalOperations <= 0) {
+            LOG.debug("No operations executed in this processor. Close immediately");
+            return false;
+        }
+        LOG.trace("Calling close on bulk processor {}", bulkProcessor);
+        bulkProcessor.close();
+        LOG.trace("Bulk Processor {} closed", bulkProcessor);
+
+        // de-register main controller
+        final int phase = phaser.arriveAndDeregister();
```

Comment:
I see it's moving code... but why "final"? I think it's not necessary here (just clutter). Again matter of taste.